tests/entropy.c

```diff
@@ -0,0 +1,189 @@
+    uint32_t dsize = 0;
+    uint32_t dcurr = 0;
+    signed int len;
+
+    do {
+	if (dsize - dcurr < BLK_SIZE) {
+	    dsize = dsize ? dsize * 2 : BLK_SIZE;
+	    data = realloc(data, dsize);
+	}
+
+	len = fread(data + dcurr, 1, BLK_SIZE, infp);
+	if (len > 0)
+	    dcurr += len;
+    } while (len > 0);
+
+    if (len == -1) {
+	perror("fread");
+    }
+
+    *lenp = dcurr;
+    return data;
+}
+
+int main(int argc, char **argv) {
+    FILE *infp = stdin;
+
+#ifdef _WIN32
+        _setmode(_fileno(stdin),  _O_BINARY);
+        _setmode(_fileno(stdout), _O_BINARY);
+#endif
+
+    extern void force_sw32_decoder(void);
+    extern void rans_disable_avx512(void);
+    extern void rans_disable_avx2(void);
+
+    if (argc > 1) {
+	if (!(infp = fopen(argv[1], "rb"))) {
+	    perror(argv[1]);
+	    return 1;
+	}
+    }
+
+    uint32_t in_size, csize, usize;
+    unsigned char *in = load(infp, &in_size);
+    int order_a[] = {0,1,                            // r4x8
+		     64,65, 128,129, 192,193,        // r4x16, arith
+		     4,5, 68,69, 132,133, 194,197,   // r4x16 SIMD
+		     };
+    char *codec[] = {"r4x8", "r4x16", "r32x16", "arith"};
+    int i, j;
+    for (i = 0; i < sizeof(order_a) / sizeof(*order_a); i++) {
+	int order = order_a[i];
+	uint8_t *comp, *uncomp;
+	for (j = 0; j < 4; j++) {
+	    int chigh = 4, clow = 0, c;
+	    for (c = 0; c < 4; c++) {
+		// Test combinations of SIMD implementations
+		uint32_t chex = (clow<<8) | chigh;
+		clow  = 1<<c;
+		chigh >>= 1;
+		rans_set_cpu(chex);
+
+		// encode
+		switch (j) {
+		case 0: // r4x8
+		    if (i >= 2) continue;
+		    comp = rans_compress(in, in_size, &csize, order);
+		    break;
+
+		case 1: // r4x16
+		    if (i >= 8) continue;
+		    comp = rans_compress_4x16(in, in_size, &csize, order);
+		    break;
+
+		case 2: // r32x16
+		    if (i < 8) continue;
+		    comp = rans_compress_4x16(in, in_size, &csize, order);
+		    break;
+
+		case 3: // arith
+		    if (i >= 8) continue;
+		    comp = arith_compress(in, in_size, &csize, order);
+		    break;
+		}
+		if (j == 2)
+		    printf("%10s-o%d-c%04x\t", codec[j], order, chex);
+		else
+		    printf("%10s-o%d      \t", codec[j], order);
+		printf("%10d uncomp, %10d comp", in_size, csize);
+
+		// decode
+		switch (j) {
+		case 0: // r4x8
+		    if (i >= 2) continue;
+		    uncomp = rans_uncompress(comp, csize, &usize);
+		    break;
+
+		case 1: // r4x16
+		    if (i >= 8) continue;
+		    uncomp = rans_uncompress_4x16(comp, csize, &usize);
+		    break;
+
+		case 2: // r32x16
+		    if (i < 8) continue;
+		    uncomp = rans_uncompress_4x16(comp, csize, &usize);
+		    break;
+
+		case 3: // arith
+		    if (i >= 8) continue;
+		    uncomp = arith_uncompress(comp, csize, &usize);
+		    break;
+		}
+
+		if (usize != in_size || memcmp(in, uncomp, usize) != 0)
+		    printf("\tFAIL\n");
+		else
+		    printf("\tpass\n");
+
+		free(comp);
+		free(uncomp);
+	    }
+	}
+	printf("\n");
+    }
+
+    free(in);
+
+    return 0;
```

Comment:
This program can't return a failure code to the test harness.

I have an improved version that can, and also check that the compressed data is the same between the alternate optimised versions of the algorithms.  Would you like it?

Reply:
Yes that's a bit of an oversight.  Thank you.

I have an amendment to it also to set stack size so we can validate it passes on e.g. Alpine linux. (It does, to be pushed as soon as I bug fix one other case.)

Reply:
Calling `pthread_exit()` also means you can't return an exit code.  Is it there simply so you can check that there is no reachable memory left using `valgrind`?  Neither `valgrind` nor Address Sanitizer seem too worried about the memory that gets left behind without doing this.

It's a bit annoying that the destructor passed to `pthread_key_create()` doesn't get called when the main thread exits, though.  It might be possible to work around by registering an `atexit()` handler, but it's a bit difficult to tell if the result would be to free the memory, or to go catastrophically wrong...